core/cont/inc/TCollection.h

```diff
@@ -272,6 +272,60 @@ inline TIter TCollection::begin() const { return ++(TIter(this)); }
 inline TIter TCollection::end() const { return TIter::End(); }
 
 
+
+/// @brief TTypedIter is a typed version of TIter.
+///
+/// The typical use is:
+/// ~~~ {.cpp}
+///    TTypedIter<TBaseClass> next(cl->GetListOfBases());
+///    while(auto bcl = next()) {
+///       ... use bcl as a TBaseClass*
+///    }
+/// ~~~ {.cpp}
+template <class Containee> // Containee must derive from TObject.
+class TTypedIter : public TIter
+{
+   static_assert(std::is_base_of<TObject, Containee>::value,
+                 "Containee type must inherit from TObject");
+
+   /// This is a workaround against ClassDefInline not supporting classes
+   /// missing their default constructor or having them private.
+   template <class T> friend class ROOT::Internal::ClassDefGenerateInitInstanceLocalInjector;
+
+   TTypedIter() = default;
+
+public:
+
+   using TIter::TIter;
+   TTypedIter(const TIter &iter) : TIter(iter) {}
+
+   Containee *operator()() { return static_cast<Containee *>(TIter::Next()); }
+   Containee *Next() { return static_cast<Containee *>(TIter::Next()); }
+   Containee *operator*() const { return static_cast<Containee *>( TIter::operator*() ); }
+
+   ClassDefInline(TTypedIter, 0);
+};
+
+/// @brief TRangeCast is an adaptater class allows the typed iteration
+/// through a TCollection.
+///
+/// The typical use is:
+/// ~~~ {.cpp}
+///    for(auto bcl : TRangeCast<TBaseClass>( *cl->GetListOfBases() )) {
```

Comment:
Would it make sense to have a `TCollection*` overload for the constructor, following the ROOT6 spirit?

Reply:
To me this would make the programming model slimmer: I would be in favour.